hpx/parallel/algorithms/detail/dispatch.hpp

```diff
@@ -19,104 +17,164 @@
+                return parallel::v1::detail::handle_exception<
+                        parallel_task_execution_policy, result_type
+                    >::call();
+            }
+        }
 
-#define BOOST_PP_ITERATION_PARAMS_1                                           \
-    (3, (2, 5, "hpx/parallel/algorithms/detail/dispatch.hpp"))                \
-    /**/
+        template <typename ExPolicy, typename... Args>
+        typename parallel::v1::detail::algorithm_result<ExPolicy, result_type>::type
+        call(ExPolicy const& policy, boost::mpl::false_, Args&&... args) const
+        {
+            return Derived::parallel(policy, std::forward<Args>(args)...);
+        }
 
-#include BOOST_PP_ITERATE()
+        ///////////////////////////////////////////////////////////////////////////
+        template <typename... Args>
+        result_type
+        call(parallel::v1::execution_policy const& policy, boost::mpl::false_,
+            Args&&... args) const
+        {
+            switch(detail::which(policy))
+            {
+            case detail::execution_policy_enum::sequential:
+                return call(*policy.get<sequential_execution_policy>(),
+                    boost::mpl::true_(), std::forward<Args>(args)...);
+
+            case detail::execution_policy_enum::sequential_task:
+                return call(seq, boost::mpl::true_(), std::forward<Args>(args)...);
+
+            case detail::execution_policy_enum::parallel:
+                return call(*policy.get<parallel_execution_policy>(),
+                    boost::mpl::false_(), std::forward<Args>(args)...);
+
+            case detail::execution_policy_enum::parallel_task:
+                {
+                    parallel_task_execution_policy const& t =
+                        *policy.get<parallel_task_execution_policy>();
+                    return call(par(t.get_executor(), t.get_chunk_size()),
+                        boost::mpl::false_(), std::forward<Args>(args)...);
+                }
+
+            case detail::execution_policy_enum::parallel_vector:
+                return call(*policy.get<parallel_vector_execution_policy>(),
+                    boost::mpl::false_(), std::forward<Args>(args)...);
+
+            default:
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    std::string("hpx::parallel::") + name_,
+                    "Not supported execution policy");
+            }
+        }
 
-        explicit algorithm(char const* const name) : name_(name) {}
+        template <typename... Args>
+        result_type
+        call(parallel::v1::execution_policy const& policy, boost::mpl::true_,
+            Args&&... args) const
+        {
+            return call(seq, boost::mpl::true_(), std::forward<Args>(args)...);
+        }
 
+    private:
         char const* const name_;
 
+        friend class boost::serialization::access;
+
         template <typename Archive>
         void serialize(Archive& ar, unsigned int)
         {
```

Comment:
The default constructors for the algorithms always set this value, no need to serialize it explicitly